src/Microsoft.OpenApi/Interfaces/IOpenApiSerializable.cs

```diff
@@ -14,7 +14,8 @@ public interface IOpenApiSerializable : IOpenApiElement
         /// Serialize Open API element to v3.0.
         /// </summary>
         /// <param name="writer">The writer.</param>
-        void SerializeAsV3(IOpenApiWriter writer);
+        /// <param name="specVersion">The OpenApi specification version.</param>
```

Comment:
What's the benefit of making it optional here? It would be miss-aligned with IOpenAPIExtension

Reply:
We don't really need the spec version when serializing most of the components in v3; we only use it to make a clear distinction when writing out  properties in v3.1 hence the need to make it optional.

Reply:
My concern here is that by making it optional we'll get undesired default behaviour or that the actual value from the parent will be forgotten to be passed from the parent, loosing some context.

Reply:
There should be a `SerializeAsV31`.   If the implementation of that method simply calls SerializeAsV3 that is fine.  The reason we created this as an interface was to guarantee that we explicitly covered every model object.

Reply:
Feel free to challenge me on this.  That was the intended design, but now is the right time to question it.

Reply:
The challenge with that approach is that it introduces a bunch of fork trees along the model.
And logic might need to be duplicated along the way. E.G. all the serialize as v2 need to call serialize as v2, all the v3 to call v3 and 31 to call 31. If one thing has a different behaviour on a leaf, I now need to stop calling serialize as v3 in 31 all along the tree for that simple leaf.

I'd be in favor of having a single serialize method, which accepts a mandatory parameter for the target version. Yes it'll require us to add a bunch of conditions but we won't have fork trees anymore.
Thoughts?

Reply:
or a better alternative would be to have the serialize as v2, v3, v31 methods for each implementation and a Serialize method at the most common ancestor that would act as a router. This way we avoid the conditions AND the fork trees.
The serialize as v2/3/31 methods from OpenAPIDocument for example would call in the Serialize (not as something) method from the router, which in turns would call the right serialize as v from the implementer.
Thoughts?

Reply:
@MaggieKimani1 calling your attention on this comment seeing your recent changes. Let me know if you want me to expand a bit on it.